src/Http/AccessToken.php

```diff
@@ -12,45 +12,78 @@
 use Carbon\Carbon;
 use Flarum\Database\AbstractModel;
 use Flarum\User\User;
+use Illuminate\Database\Eloquent\Builder;
+use Illuminate\Database\Schema\Builder as SchemaBuilder;
+use Illuminate\Support\Arr;
 use Illuminate\Support\Str;
+use Psr\Http\Message\ServerRequestInterface;
 
 /**
+ * @property int $id
  * @property string $token
  * @property int $user_id
  * @property Carbon $created_at
  * @property Carbon|null $last_activity_at
- * @property int $lifetime_seconds
+ * @property string $type
+ * @property string $title
+ * @property string $last_ip_address
+ * @property string $last_user_agent
  * @property \Flarum\User\User|null $user
  */
 class AccessToken extends AbstractModel
 {
+    protected $table = 'access_tokens';
+
+    protected $dates = [
+        'created_at',
+        'last_activity_at',
+    ];
+
     /**
-     * Use a custom primary key for this model.
+     * A map of access token types, as specified in the `type` column, to their classes.
      *
-     * @var bool
+     * @var array
      */
-    public $incrementing = false;
+    protected static $models = [];
 
-    protected $primaryKey = 'token';
+    /**
+     * The type of token this is, to be stored in the access tokens table.
+     *
+     * Should be overwritten by subclasses with the value that is
+     * to be stored in the database, which will then be used for
+     * mapping the hydrated model instance to the proper subtype.
+     *
+     * @var string
+     */
+    public static $type = '';
 
-    protected $dates = ['last_activity_at'];
+    /**
+     * How long this access token should be valid from the time of last activity.
+     * This value will be used in the validity and expiration checks.
+     * @var int Lifetime in seconds. Zero means it will never expire.
+     */
+    protected static $lifetime = 0;
```

Comment:
Shouldn't the default token type have a non-infinite lifetime?

Reply:
That's up for debate. I had no particular preference.

Reply:
Could we make the default type an abstract class so it can't be instantiated? Then it wouldn't matter. This might cause issues with the retrieval logic, I'm not sure if Laravel would automatically convert it to the proper type. It'd also require that if extensions want to introduce new types, they do so properly.

Reply:
I just tested, we cannot make it abstract. It doesn't work with Eloquent.

Again, we have the same situation with `Post` where the base post isn't abstract.

I guess the question here is whether we prefer 0 as the default, or 3600 ? I don't think any other value makes sense.

Reply:
Can we set it to null (or not have it at all) in the base class?

Reply:
Personally this seems even more confusing. Despite the lack of parameter type-hinting (which I see came in PHP 7.4 or 8?), having a number there makes it obvious what you should set.

If it's `null` and someone doesn't override it, then I don't think there will be any error, but it also won't behave differently from `0` since it's used in multiple math operations.